pkg/gardenlet/controller/networkpolicy/add.go

```diff
@@ -0,0 +1,130 @@
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/controllerutils/mapper"
+	"github.com/gardener/gardener/pkg/gardenlet/controller/networkpolicy/helper"
+	"github.com/gardener/gardener/pkg/gardenlet/controller/networkpolicy/hostnameresolver"
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "networkpolicy"
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager, seedCluster cluster.Cluster) error {
+	if r.SeedClient == nil {
+		r.SeedClient = seedCluster.GetClient()
+	}
+	if r.ShootNamespaceSelector == nil {
+		r.ShootNamespaceSelector = labels.SelectorFromSet(labels.Set{
+			v1beta1constants.GardenRole: v1beta1constants.GardenRoleShoot,
+		})
+	}
+	if r.Resolver == nil {
+		resolver, err := hostnameresolver.CreateForCluster(seedCluster.GetConfig(), mgr.GetLogger())
+		if err != nil {
+			return fmt.Errorf("failed to get hostnameresolver: %w", err)
+		}
+		r.Resolver = resolver
+	}
+
+	return builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(&corev1.Namespace{}).
+		Watches(
+			source.NewKindWithCache(&corev1.Endpoints{}, seedCluster.GetCache()),
+			mapper.EnqueueRequestsFrom(mapper.MapFunc(r.MapToNamespaces), mapper.UpdateWithNew, mgr.GetLogger()),
+			builder.WithPredicates(r.IsKubernetesEndpoint()),
+		).
+		Watches(
+			source.NewKindWithCache(&networkingv1.NetworkPolicy{}, seedCluster.GetCache()),
+			mapper.EnqueueRequestsFrom(mapper.MapFunc(r.MapObjectToNamespace), mapper.UpdateWithNew, mgr.GetLogger()),
+			builder.WithPredicates(r.IsAllowToSeedApiserverNetworkPolicy()),
+		).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: *r.Config.ConcurrentSyncs,
+			RecoverPanic:            true,
+			RateLimiter:             workqueue.DefaultControllerRateLimiter(),
+		}).
+		Complete(r)
+}
+
+// MapToNamespaces is a mapper function which returns requests for all shoot namespaces + garden namespace + istio-system namespace.
+func (r *Reconciler) MapToNamespaces(ctx context.Context, log logr.Logger, _ client.Reader, _ client.Object) []reconcile.Request {
+	namespaces := &corev1.NamespaceList{}
+	if err := r.SeedClient.List(ctx, namespaces, &client.ListOptions{
+		LabelSelector: r.ShootNamespaceSelector,
+	}); err != nil {
+		log.Error(err, "Unable to list Shoot namespace for updating NetworkPolicy", "networkPolicyName", helper.AllowToSeedAPIServer)
+		return []reconcile.Request{}
+	}
+
+	requests := []reconcile.Request{}
+	for _, namespace := range namespaces.Items {
+		requests = append(requests, reconcile.Request{NamespacedName: client.ObjectKeyFromObject(&namespace)})
+	}
+	requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: v1beta1constants.GardenNamespace}})
+	requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: v1beta1constants.IstioSystemNamespace}})
```

Comment:
Instead of adding those here, you could initialise the `requests` in L95 directly.